pubsub/google/cloud/pubsub_v1/subscriber/_protocol/bidi.py

```diff
@@ -0,0 +1,489 @@
+                yield self._initial_request()
+            else:
+                yield self._initial_request
+
+        while True:
+            try:
+                item = self._queue.get(timeout=self._period)
+            except queue.Empty:
+                if not self._is_active():
+                    _LOGGER.debug(
+                        'Empty queue and inactive call, exiting request '
+                        'generator.')
+                    return
+                else:
+                    # call is still active, keep waiting for queue items.
+                    continue
+
+            # The consumer explicitly sent "None", indicating that the request
+            # should end.
+            if item is None:
+                _LOGGER.debug('Cleanly exiting request generator.')
+                return
+
+            if not self._is_active():
+                # We have an item, but the call is closed. We should put the
+                # item back on the queue so that the next call can consume it.
+                self._queue.put(item)
+                _LOGGER.debug(
+                    'Inactive call, replacing item on queue and exiting '
+                    'request generator.')
+                return
+
+            yield item
+
+
+class BidiRpc(object):
+    """A helper for consuming a bi-directional streaming RPC.
+
+    This maps gRPC's built-in interface which uses a request iterator and a
+    response iterator into a socket-like :func:`send` and :func:`recv`. This
+    is a more useful pattern for long-running or asymmetric streams (streams
+    where there is not a direct correlation between the requests and
+    responses).
+
+    Example::
+
+        initial_request = example_pb2.StreamingRpcRequest(
+            setting='example')
+        rpc = BidiRpc(stub.StreamingRpc, initial_request=initial_request)
+
+        rpc.open()
+
+        while rpc.is_active():
+            print(rpc.recv())
+            rpc.send(example_pb2.StreamingRpcRequest(
+                data='example'))
+
+    This does *not* retry the stream on errors. See :class:`ResumableBidiRpc`.
+
+    Args:
+        start_rpc (grpc.StreamStreamMultiCallable): The gRPC method used to
+            start the RPC.
+        initial_request (Union[protobuf.Message,
+                Callable[None, protobuf.Message]]): The initial request to
+            yield. This is useful if an initial request is needed to start the
+            stream.
+    """
+    def __init__(self, start_rpc, initial_request=None):
+        self._start_rpc = start_rpc
+        self._initial_request = initial_request
+        self._request_queue = queue.Queue()
+        self._request_generator = None
+        self._is_active = False
+        self.call = None
+        self._callbacks = []
+
+    def add_done_callback(self, callback):
+        """Adds a callback that will be called when the RPC terminates.
+
+        This occurs when the RPC errors or is successfully terminated.
```

Comment:
Is it possible to split these?

Reply:
No, this matches the future interface that combines both.